### PR TITLE
Reuse binaries for tuist test, generate test project in the current working directory

### PR DESCRIPTION
**Pull request by fortmarek**
Resolves https://github.com/tuist/tuist/issues/5281
Resolves https://github.com/tuist/tuist/issues/5184
Resolves https://github.com/tuist/tuist/issues/5034

### Short description 📝

This PR does two things:
- `tuist test` generates the project in-place. That resolves issues #5281 and #5184. Generating project in the custom `~/.tuist/Cache/Projects` directory was causing confusion. While we will now re-generate projects for the purposes of `tuist test`, this can actually be a plus. If tests fail, developers can easily open the generated project and retest their code using the Xcode UI while having the project lean.
- `tuist test` will now reuse the binaries generated by `tuist cache warm` which will make the tests a lot faster as developers and the CI will no longer need to rebuild modules that have been pre-warmed. **Note:** This functionality is available only to users of Tuist Cloud.

### How to test the changes locally 🧐

Run `tuist test` for a project that has tests and Tuist Cloud. You can use `tuist/tuist` as well. The generated project should use binaries and only unit tests should be integrated as sources.

### Contributor checklist ✅

- [x] The code has been linted using run `make lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Comment by JohnnyTseng** [comment]
Hi, 

I've been experimenting `tuist cache warm` recently, but found out it seems like `xcodebuild` still needs to build external SPM dependencies, if I am correct, this is currently an "expected behavior" and this PR addresses the issue I have?

Thank you!

**Comment by fortmarek** [comment]
> I've been experimenting tuist cache warm recently, but found out it seems like xcodebuild still needs to build external SPM dependencies, if I am correct, this is currently an "expected behavior" and this PR addresses the issue I have?

Depends on what you are invoking. But yes, if you use `tuist test` and prewarmed your cache, you shouldn't see any build logs from `xcodebuild` concerning external dependencies (and any other internal targets that are in the cache)

File: Sources/TuistKit/Commands/TestCommand.swift [hunk 1]
```diff
@@ -102,6 +108,37 @@ struct TestCommand: AsyncParsableCommand {
     )
     var skipConfigurations: [String] = []
 
+    @Flag(
+        name: [.customShort("x"), .long],
+        help: "When passed it uses xcframeworks (simulator and device) from the cache instead of frameworks (only simulator)."
+    )
+    var xcframeworks: Bool = false
```
Comment (danyf90) [inline Sources/TuistKit/Commands/TestCommand.swift:115]: Maybe we can re-use
```
        @OptionGroup()
        var options: CacheOptions
```
?

**Review by danyf90 — APPROVED** [review]

**Comment by JohnnyTseng** [comment]
> > I've been experimenting tuist cache warm recently, but found out it seems like xcodebuild still needs to build external SPM dependencies, if I am correct, this is currently an "expected behavior" and this PR addresses the issue I have?
> 
> Depends on what you are invoking. But yes, if you use `tuist test` and prewarmed your cache, you shouldn't see any build logs from `xcodebuild` concerning external dependencies (and any other internal targets that are in the cache)

Thanks @fortmarek !
If I understood correctly, running `tuist cache warm` only gets benefit when running command `tuist test` & `tuist build` on Tuist Cloud, is that correct?

File: Sources/TuistKit/Commands/TestCommand.swift (hunk 1 above)
Comment (pepicrft) [reply Sources/TuistKit/Commands/TestCommand.swift:115]: Agreed! Consistency is important across commands.

**Review by pepicrft — APPROVED** [review]

**Comment by fortmarek** [comment]
>  If I understood correctly, running tuist cache warm only gets benefit from cache when running command tuist test / tuist build on Tuist Cloud, is that correct?

`tuist build` does not reuse the cache artifacts just yet, but it will soon. Additionally, `tuist generate` also reuses the artifacts, so you can develop in a lean Xcode project 🙂 You can also run `tuist generate MyCurrentTarget` which will take binaries for all cached, internal targets that `MyCurrentTarget` depends on.

File: Sources/TuistKit/Commands/TestCommand.swift (hunk 1 above)
Comment (fortmarek) [reply Sources/TuistKit/Commands/TestCommand.swift:115]: We want to have different descriptions of those flags

**Pull request merged**